File: apps/ios/MiPlancha/Waiter/TablesService.swift
//
//  TablesService.swift
//  MiPlancha
//
//  Created by julio on 25/08/2026.
//

import Observation
import FirebaseFirestore
import FirebaseAuth

@Observable
@MainActor
final class TablesService {
    private var listener: ListenerRegistration?
    private(set) var tables: [Table] = []
    private var companyId = ""
    private(set) var clientNames: [String : String] = [:]
    private(set) var clientSatAt: [String : Date] = [:]
    private(set) var tableOrderInfo: [Int: TableOrderSummary] = [:]
    private var listenerLines: ListenerRegistration?
    
    func startListening(companyId: String) {
        self.companyId = companyId
        
        listener = Firestore.firestore()
            .collection("empresas").document(companyId).collection("mesas")
            .addSnapshotListener { [weak self] snapshot, _ in
                guard let docs = snapshot?.documents else { return }
                let parsed = docs.compactMap { doc -> Table? in
                    let data = doc.data()
                    let clienteId = data["clienteId"] as? String
                    guard let number = data["numero"] as? Int,
                          let statusRaw = data["estado"] as? String,
                          let status = TableStatus(rawValue: statusRaw)
                    else { return nil }
                    return Table(id: doc.documentID, number: number, status: status, clientId: clienteId)
                }
                Task { @MainActor [weak self] in
                    guard let self else { return }
                    self.tables = parsed
                    
                    for table in parsed {
                        guard let clientId = table.clientId,
                              self.clientNames[table.id] == nil else {continue}
                        
                        let doc = try? await Firestore.firestore()
                            .collection("empresas").document(self.companyId)
                            .collection("clientes").document(clientId)
                            .getDocument()
                        
                        if let name = doc?.data()?["nombre"] as? String{
                            self.clientNames[table.id] = name
                        }
                        
                        if let ts = doc?.data()?["abiertoEn"] as? Timestamp {
                            self.clientSatAt[table.id] = ts.dateValue()
                        }
                    }
                    
                    let occupiedIds = Set(parsed.filter { $0.status == .ocupada }.map { $0.id })
                    clientNames = clientNames.filter { occupiedIds.contains($0.key) }
                    clientSatAt = clientSatAt.filter { occupiedIds.contains($0.key) }
                }
            }
        
        listenerLines = Firestore.firestore()
            .collectionGroup("lineas")
            .whereField("empresaId", isEqualTo: companyId)
            .whereField("estado", isNotEqualTo: LineStatus.ready.rawValue)
            .addSnapshotListener { snapshot, _ in
                guard let docs = snapshot?.documents else {return }
                
                var grouped: [Int:[LineStatus]] = [:]
                var timestamps: [Int:Date] = [:]
                
                for doc in docs {
                    let data = doc.data()
                    guard let tableNumber = data["mesaNumero"] as? Int,
                          let rawStatus = data["estado"] as? String,
                          let status = LineStatus(rawValue: rawStatus) else { return}
                    
                    grouped[tableNumber, default: []].append(status)
                    
                    let ts = (data["colocadoEn"] as? Timestamp)?.dateValue()
                    ?? (data["pedidoCreadoEn"] as? Timestamp)?.dateValue()
                    if let ts {
                        let current = timestamps[tableNumber]
                        timestamps[tableNumber] = current.map { max($0, ts) } ?? ts
                    }
                }
                
                print("voy a mostrar info de mesas")
                Task { @MainActor [weak self] in
                    self?.tableOrderInfo = grouped.compactMapValues { statuses in
                        print("STATUS: \(statuses)")
                        
                        guard let worst = statuses.min(by: { Self.priority($0) < Self.priority($1) }),
                              let date = timestamps[...].first?.value else { return nil }
                        
                        print("WORST: \(worst) - DATE: \(date)")
                        return TableOrderSummary(worstStatus: worst, lastUpdate: date)
                    }
                }
            }
    }
    
    func stopListening() {
        listener?.remove()
        listener = nil
        
        listenerLines?.remove()
        listenerLines = nil
    }
    
    func openTable(_ table: Table, clientName: String) async throws {
        guard let uid = Auth.auth().currentUser?.uid else {return}
        
        let db = Firestore.firestore()
        let clientRef = db.collection("empresas").document( companyId)
            .collection("clientes").document()
        
        try await clientRef.setData([
            "camareroId": uid,
            "mesaId": table.id,
            "nombre": clientName,
            "abiertoEn": FieldValue.serverTimestamp()
        ])
        
        try await db.collection("empresas").document(companyId)
            .collection("mesas").document(table.id)
            .updateData(["estado": "ocupada", "clienteId": clientRef.documentID])
        
        
    }
    
    func closeTable(_ table: Table) async throws {
        try await Firestore.firestore()
            .collection("empresas").document(companyId)
            .collection("mesas").document(table.id)
            .updateData(["estado": "libre", "clienteId": NSNull()])
    }
    
    private static func priority(_ status: LineStatus) -> Int {
        switch status {
        case .pending: return 0
        case .cooking: return 1
        case .pendingDelivery: return 2
        default: return 3
        }
    }
}

struct TableOrderSummary {
    let worstStatus: LineStatus
    let lastUpdate: Date
}

File: apps/android/app/src/main/java/com/jugomo/miplancha/waiter/TablesService.kt
package com.jugomo.miplancha.waiter

import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await
import androidx.compose.runtime.mutableStateMapOf
import com.google.firebase.Timestamp
import com.google.firebase.firestore.FieldValue

class TablesService {

    private val db = FirebaseFirestore.getInstance()
    private var companyId: String = ""
    var clientsCache = mutableStateMapOf<String, Client>()
    var tableOrderInfo = mutableStateMapOf<Int, TableOrderSummary>()
    var tablesAllDelivered = mutableStateMapOf<Int, Timestamp>()

    fun startListeningTables(companyId: String): Flow<List<Table>?> {
        this@TablesService.companyId = companyId

        return callbackFlow {
            val docRef = db.collection("empresas")
                .document(companyId)
                .collection("mesas")

            val listenerRegistration = docRef.addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }

                if (snapshot != null) {
                    val mesas = snapshot.documents.map { it.toTable() }
                    cleanCacheClient(mesas)
                    trySend(mesas)
                } else {
                    trySend(null)
                }
            }

            awaitClose {
                listenerRegistration.remove()
            }
        }
    }

    fun startListeningOrdersByTable(companyId: String): Flow<Map<Int, List<LineStatus>>?> {
        return callbackFlow {
            val docRef = db.collectionGroup("lineas")
                .whereEqualTo("empresaId", companyId)
                .whereNotEqualTo("estado", "listo")

            val listenerRegistration = docRef.addSnapshotListener { snapshot, error ->
                if(error != null) {
                    close(error)
                    return@addSnapshotListener
                }

                if (snapshot != null) {
                    val porMesa: Map<Int, List<LineStatus>> =
                        snapshot.documents
                            .mapNotNull { doc ->
                                val mesa = doc.getLong("mesaNumero")?.toInt() ?:
                                return@mapNotNull null
                                val estado = doc.getString("estado") ?:
                                return@mapNotNull null
                                mesa to LineStatus.valueOf(estado.uppercase())
                            }
                            .groupBy({ it.first }, { it.second })

                    val previas = tableOrderInfo.keys.toSet()

                    porMesa.forEach { (numeroMesa, estados) ->
                        val peor = worstStatus(estados) ?: return@forEach
                        tableOrderInfo[numeroMesa] = TableOrderSummary(
                            worstStatus = peor,
                            lastUpdate = Timestamp.now(),
                            displayLabel = null
                        )
                    }

                    val yaEntregadas = previas - porMesa.keys
                    yaEntregadas.forEach { numeroMesa ->
                        tablesAllDelivered[numeroMesa] = Timestamp.now()
                        tableOrderInfo.remove(numeroMesa)
                    }

                    trySend(porMesa)
                } else {
                    trySend(null)
                }
            }

            awaitClose { listenerRegistration.remove() }
        }
    }

    suspend fun openTable(table: Table, waiterId: String, clientName: String) {
        val tableRef = db.collection("empresas")
            .document(companyId)
            .collection("mesas")
            .document(table.id)
        val clientRef = db.collection("empresas")
            .document(companyId)
            .collection("clientes")
            .document()

        val docRef = db.runTransaction { trn ->
            val tableSnp = trn.get(tableRef)
            if(tableSnp.exists() && tableSnp.data!!["estado"] == "libre") {
                trn.set(
                    clientRef,
                    mapOf(
                        "mesaId" to table.id,
                        "nombre" to clientName,
                        "camareroId" to waiterId,
                        "abiertoEn" to FieldValue.serverTimestamp()
                    )
                )
                trn.update(
                    tableRef,
                    mapOf(
                        "estado" to "ocupada",
                        "clienteId" to clientRef.id
                    )
                )
            } else {
                throw MesaOcupadaError(msg = "no se pudo abrir la mesa")
            }
        }.await()
    }

    suspend fun backfillDeliveredIfNeeded(tableNumber: Int, clientSatAt: Timestamp?) {
        val snap = db.collectionGroup("lineas")
            .whereEqualTo("empresaId", companyId)
            .whereEqualTo("mesaNumero", tableNumber)
            .whereEqualTo("estado", "listo")
            .get()
            .await()

        val dates = snap.documents.mapNotNull { doc ->
            val createdAt = doc.getTimestamp("pedidoCreadoEn") ?: return@mapNotNull null
            if (clientSatAt != null  && createdAt < clientSatAt) return@mapNotNull null
            doc.getTimestamp("colocadoEn" ) ?: createdAt
        }

        val last = dates.maxOrNull() ?: return

        if(tableOrderInfo[tableNumber] == null) {
            tablesAllDelivered[tableNumber] = last
        }
    }

    fun orderSummary(
        table: Table,
        tableOrderInfo : Map<Int, TableOrderSummary>,
        tablesAllDelivered: Map<Int, Timestamp>,
        clientSatAt: Timestamp?
    ) : TableOrderSummary?{
        tableOrderInfo[table.numero]?.let { return it }

        tablesAllDelivered[table.numero]?.let {
            return TableOrderSummary(
                worstStatus = LineStatus.LISTO,
                lastUpdate = it,
                displayLabel = null
            )
        }

        if(clientSatAt != null) {
            return TableOrderSummary(
                worstStatus = LineStatus.PENDIENTE,
                lastUpdate = clientSatAt,
                displayLabel = "Esperando camarero"
            )
        }

        return null
    }


    suspend fun getClient(clientId: String): Client? {
        val docRef = db.collection("empresas")
            .document(companyId)
            .collection("clientes")
            .document(clientId)

        val snapshot = docRef.get().await()
        return if (snapshot.exists()) snapshot.toClient() else null

    }

    suspend fun cacheClient(table: Table) {
        val clientId = table.clienteId ?: return

        if(clientId !in clientsCache ){
            clientsCache[clientId] = getClient(clientId) ?: return
        }
    }

    fun cleanCacheClient(tables: List<Table>) {
        clientsCache.keys.retainAll(tables.mapNotNull { it.clienteId }.toSet())
    }
}

class MesaOcupadaError(msg: String) : Exception(msg) {
}